File: be/routes/repairs.js
const express = require('express');
const { RepairTicket, RepairItem, Vehicle } = require('../models');
const { authMiddleware, requireRole } = require('../middleware/auth');
const { sendSuccess, sendError } = require('../utils/response');
const statusCodes = require('../constants/statusCodes');
const errorMessages = require('../constants/errorMessages');
const { ROLES } = require('../constants/roles');
const logger = require('../utils/logger');

const router = express.Router();
router.use(authMiddleware);

// GET /api/repairs/my-tasks - Get repairs assigned to the logged-in mechanic
router.get('/my-tasks', requireRole(ROLES.ADMIN, ROLES.MECHANIC), async (req, res) => {
  try {
    const { fullName } = req.user;

    const tickets = await RepairTicket.findAll({
      where: { mechanicName: fullName },
      include: [
        { model: Vehicle, as: 'vehicle' },
        { model: RepairItem, as: 'items' }
      ],
      order: [['createdAt', 'DESC']]
    });

    sendSuccess(res, tickets);
  } catch (error) {
    logger.error('Error fetching mechanic tasks:', error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

// GET /api/repairs/my-repairs - Get repairs for vehicles belonging to logged-in customer
router.get('/my-repairs', async (req, res) => {
  try {
    const { fullName } = req.user;

    // Find all vehicles belonging to this customer
    const vehicles = await Vehicle.findAll({
      where: { customerName: fullName },
      attributes: ['id']
    });

    const vehicleIds = vehicles.map(v => v.id);

    if (vehicleIds.length === 0) {
      return sendSuccess(res, []);
    }

    // Find all repair tickets for those vehicles
    const tickets = await RepairTicket.findAll({
      where: { vehicleId: vehicleIds },
      include: [
        { model: Vehicle, as: 'vehicle' },
        { model: RepairItem, as: 'items' }
      ],
      order: [['createdAt', 'DESC']]
    });

    sendSuccess(res, tickets);
  } catch (error) {
    logger.error('Error fetching customer repairs:', error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

// GET /api/repairs - Get all repair tickets (Admin only)
router.get('/', requireRole(ROLES.ADMIN), async (req, res) => {
  try {
    const tickets = await RepairTicket.findAll({
      include: [
        { model: Vehicle, as: 'vehicle' },
        { model: RepairItem, as: 'items' }
      ],
      order: [['createdAt', 'DESC']]
    });

    sendSuccess(res, tickets);
  } catch (error) {
    logger.error('Error fetching repair tickets:', error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

// GET /api/repairs/:id - Get single repair ticket (Admin + Mechanic)
router.get('/:id', requireRole(ROLES.ADMIN, ROLES.MECHANIC, ROLES.CUSTOMER), async (req, res) => {
  try {
    const ticket = await RepairTicket.findByPk(req.params.id, {
      include: [
        { model: Vehicle, as: 'vehicle' },
        { model: RepairItem, as: 'items' }
      ]
    });

    if (!ticket) {
      return sendError(res, errorMessages.NOT_FOUND_REPAIR, statusCodes.NOT_FOUND);
    }

    // Customer can only view their own vehicle's repairs
    if (req.user.role === ROLES.CUSTOMER) {
      const vehicle = ticket.vehicle;
      if (!vehicle || vehicle.customerName !== req.user.fullName) {
        return sendError(res, errorMessages.AUTH_UNAUTHORIZED, statusCodes.FORBIDDEN);
      }
    }

    // Mechanic can only view repairs assigned to them
    if (req.user.role === ROLES.MECHANIC) {
      if (ticket.mechanicName !== req.user.fullName) {
        return sendError(res, errorMessages.AUTH_UNAUTHORIZED, statusCodes.FORBIDDEN);
      }
    }

    sendSuccess(res, ticket);
  } catch (error) {
    logger.error(`Error fetching repair ticket ${req.params.id}:`, error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

// POST /api/repairs - Create repair ticket (Admin only)
router.post('/', requireRole(ROLES.ADMIN), async (req, res) => {
  try {
    const { vehicleId, items, mechanicName } = req.body;

    if (!vehicleId) {
      return sendError(res, errorMessages.VALIDATION_REQUIRED_FIELD('vehicleId'), statusCodes.BAD_REQUEST);
    }

    // Calculate total
    let totalAmount = 0;
    if (items && items.length > 0) {
      totalAmount = items.reduce((sum, item) => sum + parseFloat(item.totalPrice), 0);
    }

    // Create ticket
    const ticket = await RepairTicket.create({
      vehicleId,
      totalAmount,
      status: 'draft',
      mechanicName: mechanicName || 'Chưa phân công'
    });

    // Create items
    if (items && items.length > 0) {
      const itemsData = items.map(item => ({
        ...item,
        repairTicketId: ticket.id
      }));
      await RepairItem.bulkCreate(itemsData);
    }

    // Update vehicle status to working
    const vehicle = await Vehicle.findByPk(vehicleId);
    if (vehicle && vehicle.status === 'pending') {
      await vehicle.update({ status: 'working' });
    }

    // Fetch complete ticket
    const completeTicket = await RepairTicket.findByPk(ticket.id, {
      include: [
        { model: Vehicle, as: 'vehicle' },
        { model: RepairItem, as: 'items' }
      ]
    });

    sendSuccess(res, completeTicket, 'Tạo phiếu sửa chữa thành công', statusCodes.CREATED);
  } catch (error) {
    logger.error('Error creating repair ticket:', error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

// PUT /api/repairs/:id - Update repair ticket (Admin + Mechanic)
router.put('/:id', requireRole(ROLES.ADMIN, ROLES.MECHANIC), async (req, res) => {
  try {
    const ticket = await RepairTicket.findByPk(req.params.id, {
      include: [{ model: RepairItem, as: 'items' }]
    });

    if (!ticket) {
      return sendError(res, errorMessages.NOT_FOUND_REPAIR, statusCodes.NOT_FOUND);
    }

    // Mechanic can only update repairs assigned to them
    if (req.user.role === ROLES.MECHANIC) {
      if (ticket.mechanicName !== req.user.fullName) {
        return sendError(res, 'Bạn không được phân công phiếu này', statusCodes.FORBIDDEN);
      }
      // Mechanic can only change status to 'completed'
      if (req.body.status && req.body.status !== 'completed') {
        return sendError(res, 'Thợ chỉ có thể đánh dấu hoàn thành phiếu sửa', statusCodes.FORBIDDEN);
      }
    }

    const newStatus = req.body.status;
    const oldStatus = ticket.status;

    // Validation: Check if status can transition
    if (newStatus && newStatus !== oldStatus) {
      
      // Validate completed status: all items must be completed
      if (newStatus === 'completed') {
        const hasIncompleteItems = ticket.items.some(item => !item.isCompleted);
        if (hasIncompleteItems) {
          return sendError(res, 'Không thể hoàn thành! Vẫn còn hạng mục chưa hoàn thành.', statusCodes.BAD_REQUEST);
        }
      }

      // Auto-set timestamps based on status transitions
      if (newStatus === 'working' && !ticket.startedAt) req.body.startedAt = new Date();
      if (newStatus === 'completed' && !ticket.completedAt) req.body.completedAt = new Date();
      if (newStatus === 'paid' && !ticket.paidAt) req.body.paidAt = new Date();
    }

    // Update ticket
    await ticket.update(req.body);

    // Sync vehicle status if repair status changed
    if (newStatus && newStatus !== oldStatus && ticket.vehicleId) {
      const vehicle = await Vehicle.findByPk(ticket.vehicleId);
      if (vehicle) {
        let vehicleStatus = vehicle.status;
        
        // Map repair status to vehicle status
        if (newStatus === 'draft' || newStatus === 'working') {
          vehicleStatus = 'repairing';
        } else if (newStatus === 'completed' || newStatus === 'paid') {
          vehicleStatus = 'completed';
        }
        
        // Update vehicle if status changed
        if (vehicleStatus !== vehicle.status) {
          await vehicle.update({ status: vehicleStatus });
        }
      }
    }

    // Fetch updated ticket with relations
    const updatedTicket = await RepairTicket.findByPk(ticket.id, {
      include: [
        { model: Vehicle, as: 'vehicle' },
        { model: RepairItem, as: 'items' }
      ]
    });

    sendSuccess(res, updatedTicket, 'Cập nhật thành công');
  } catch (error) {
    logger.error(`Error updating repair ticket ${req.params.id}:`, error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

// DELETE /api/repairs/:id - Delete repair ticket (Admin only)
router.delete('/:id', requireRole(ROLES.ADMIN), async (req, res) => {
  try {
    const ticket = await RepairTicket.findByPk(req.params.id);

    if (!ticket) {
      return sendError(res, errorMessages.NOT_FOUND_REPAIR, statusCodes.NOT_FOUND);
    }

    // Prevent deletion of completed/paid repairs
    if (ticket.status === 'completed' || ticket.status === 'paid') {
      return sendError(res, 'Không thể xóa phiếu đã hoàn thành hoặc đã thanh toán!', statusCodes.BAD_REQUEST);
    }

    await ticket.destroy();

    sendSuccess(res, null, 'Xóa thành công');
  } catch (error) {
    logger.error(`Error deleting repair ticket ${req.params.id}:`, error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

// PUT /api/repairs/:id/items/:itemId/toggle - Toggle repair item completion (Admin + Mechanic)
router.put('/:id/items/:itemId/toggle', requireRole(ROLES.ADMIN, ROLES.MECHANIC), async (req, res) => {
  try {
    // If mechanic, verify assignment
    if (req.user.role === ROLES.MECHANIC) {
      const ticket = await RepairTicket.findByPk(req.params.id);
      if (!ticket || ticket.mechanicName !== req.user.fullName) {
        return sendError(res, 'Bạn không được phân công phiếu này', statusCodes.FORBIDDEN);
      }
    }

    const item = await RepairItem.findOne({
      where: {
        id: req.params.itemId,
        repairTicketId: req.params.id
      }
    });

    if (!item) {
      return sendError(res, errorMessages.NOT_FOUND_INVENTORY, statusCodes.NOT_FOUND);
    }

    const isCompleted = req.body.isCompleted;

    await item.update({
      isCompleted: isCompleted,
      completedAt: isCompleted ? new Date() : null
    });

    sendSuccess(res, item, 'Đã cập nhật trạng thái hạng mục');
  } catch (error) {
    logger.error('Error toggling repair item:', error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

// GET /api/repairs/:id/can-complete - Check if repair can be completed (Admin + Mechanic)
router.get('/:id/can-complete', requireRole(ROLES.ADMIN, ROLES.MECHANIC), async (req, res) => {
  try {
    const ticket = await RepairTicket.findByPk(req.params.id, {
      include: [{ model: RepairItem, as: 'items' }]
    });

    if (!ticket) {
      return sendError(res, errorMessages.NOT_FOUND_REPAIR, statusCodes.NOT_FOUND);
    }

    const totalItems = ticket.items.length;
    const completedItems = ticket.items.filter(item => item.isCompleted).length;
    const canComplete = totalItems > 0 && completedItems === totalItems;

    sendSuccess(res, {
      canComplete,
      totalItems,
      completedItems,
      incompleteItems: totalItems - completedItems
    });
  } catch (error) {
    logger.error('Error checking repair status:', error);
    sendError(res, errorMessages.SERVER_ERROR, statusCodes.INTERNAL_SERVER_ERROR, error.message);
  }
});

module.exports = router;

File: be/routers/repairs.py
from datetime import datetime, timezone
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from database.session import get_db
from models.repair_ticket import RepairTicket
from models.repair_item import RepairItem
from models.vehicle import Vehicle
from schemas.repair import (
    RepairTicketCreate,
    RepairTicketUpdate,
    RepairTicketResponse,
    RepairItemToggle,
)
from core.constants import Role
from core.response import success_response, error_response
from middleware.auth import get_current_user, require_role

router = APIRouter(prefix="/api/repairs", tags=["Repairs"])


def _serialize_ticket(ticket: RepairTicket) -> dict:
    return RepairTicketResponse.model_validate(ticket).model_dump()


# ── Mechanic view: tasks assigned to me ───────────────────────────────────────

@router.get("/my-tasks", summary="Get repairs assigned to the logged-in mechanic")
async def my_tasks(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(require_role(Role.ADMIN, Role.MECHANIC)),
):
    stmt = (
        select(RepairTicket)
        .where(RepairTicket.mechanicName == current_user["fullName"])
        .order_by(RepairTicket.createdAt.desc())
    )
    result = await db.execute(stmt)
    tickets = result.scalars().all()
    return success_response([_serialize_ticket(t) for t in tickets])


# ── Customer view: repairs for my vehicles ────────────────────────────────────

@router.get("/my-repairs", summary="Get repair history for the logged-in customer's vehicles")
async def my_repairs(
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    # Find vehicles owned by this customer
    v_stmt = select(Vehicle.id).where(Vehicle.customerName == current_user["fullName"])
    v_result = await db.execute(v_stmt)
    vehicle_ids = [row[0] for row in v_result.all()]

    if not vehicle_ids:
        return success_response([])

    stmt = (
        select(RepairTicket)
        .where(RepairTicket.vehicleId.in_(vehicle_ids))
        .order_by(RepairTicket.createdAt.desc())
    )
    result = await db.execute(stmt)
    tickets = result.scalars().all()
    return success_response([_serialize_ticket(t) for t in tickets])


# ── Admin view: all tickets ────────────────────────────────────────────────────

@router.get("", summary="List all repair tickets (admin only)")
async def list_repairs(
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(require_role(Role.ADMIN)),
):
    stmt = select(RepairTicket).order_by(RepairTicket.createdAt.desc())
    result = await db.execute(stmt)
    tickets = result.scalars().all()
    return success_response([_serialize_ticket(t) for t in tickets])


# ── Single ticket ──────────────────────────────────────────────────────────────

@router.get("/{ticket_id}", summary="Get a single repair ticket")
async def get_repair(
    ticket_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(require_role(Role.ADMIN, Role.MECHANIC, Role.CUSTOMER)),
):
    stmt = select(RepairTicket).where(RepairTicket.id == ticket_id)
    result = await db.execute(stmt)
    ticket = result.scalar_one_or_none()

    if not ticket:
        return error_response("Không tìm thấy phiếu sửa chữa", 404)

    # Customer: only see repairs for their own vehicles
    if current_user["role"] == Role.CUSTOMER.value:
        v_stmt = select(Vehicle).where(
            Vehicle.id == ticket.vehicleId,
            Vehicle.customerName == current_user["fullName"],
        )
        v_result = await db.execute(v_stmt)
        if not v_result.scalar_one_or_none():
            return error_response("Bạn không có quyền xem phiếu này", 403)

    # Mechanic: only see repairs assigned to them
    if current_user["role"] == Role.MECHANIC.value:
        if ticket.mechanicName != current_user["fullName"]:
            return error_response("Bạn không được phân công phiếu này", 403)

    return success_response(_serialize_ticket(ticket))


# ── Create ─────────────────────────────────────────────────────────────────────

@router.post("", summary="Create a repair ticket (admin only)")
async def create_repair(
    body: RepairTicketCreate,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(require_role(Role.ADMIN)),
):
    total = sum(float(i.totalPrice) for i in (body.items or []))

    ticket = RepairTicket(
        vehicleId=body.vehicleId,
        mechanicName=body.mechanicName or "Chưa phân công",
        totalAmount=total,
        status="draft",
    )
    db.add(ticket)
    await db.flush()  # get ticket.id without full commit

    for item_data in (body.items or []):
        item = RepairItem(repairTicketId=ticket.id, **item_data.model_dump())
        db.add(item)

    # Update vehicle status to repairing
    v_stmt = select(Vehicle).where(Vehicle.id == body.vehicleId)
    v_result = await db.execute(v_stmt)
    vehicle = v_result.scalar_one_or_none()
    if vehicle and vehicle.status == "waiting":
        vehicle.status = "repairing"

    await db.commit()
    await db.refresh(ticket)
    return success_response(_serialize_ticket(ticket), "Tạo phiếu sửa chữa thành công", 201)


# ── Update ─────────────────────────────────────────────────────────────────────

@router.put("/{ticket_id}", summary="Update repair ticket status or details (admin + mechanic)")
async def update_repair(
    ticket_id: int,
    body: RepairTicketUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(require_role(Role.ADMIN, Role.MECHANIC)),
):
    stmt = select(RepairTicket).where(RepairTicket.id == ticket_id)
    result = await db.execute(stmt)
    ticket = result.scalar_one_or_none()
    if not ticket:
        return error_response("Không tìm thấy phiếu sửa chữa", 404)

    # Mechanic restrictions
    if current_user["role"] == Role.MECHANIC.value:
        if ticket.mechanicName != current_user["fullName"]:
            return error_response("Bạn không được phân công phiếu này", 403)
        if body.status and body.status != "completed":
            return error_response("Thợ chỉ có thể đánh dấu hoàn thành phiếu sửa", 403)

    new_status = body.status
    old_status = ticket.status

    if new_status and new_status != old_status:
        # All items must be done before completing
        if new_status == "completed":
            incomplete = [i for i in ticket.items if not i.isCompleted]
            if incomplete:
                return error_response("Không thể hoàn thành! Vẫn còn hạng mục chưa hoàn thành.", 400)

        # Auto-set timestamps on status transitions
        now = datetime.now(timezone.utc)
        if new_status == "working" and not ticket.startedAt:
            ticket.startedAt = now
        if new_status == "completed" and not ticket.completedAt:
            ticket.completedAt = now
        if new_status == "paid" and not ticket.paidAt:
            ticket.paidAt = now

    # Apply update fields
    for field, value in body.model_dump(exclude_none=True).items():
        setattr(ticket, field, value)

    # Sync vehicle status
    if new_status and new_status != old_status and ticket.vehicleId:
        v_stmt = select(Vehicle).where(Vehicle.id == ticket.vehicleId)
        v_result = await db.execute(v_stmt)
        vehicle = v_result.scalar_one_or_none()
        if vehicle:
            vehicle.status = "repairing" if new_status in ("draft", "working") else "completed"

    await db.commit()
    await db.refresh(ticket)
    return success_response(_serialize_ticket(ticket), "Cập nhật thành công")


# ── Delete ─────────────────────────────────────────────────────────────────────

@router.delete("/{ticket_id}", summary="Delete a repair ticket (admin only)")
async def delete_repair(
    ticket_id: int,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(require_role(Role.ADMIN)),
):
    stmt = select(RepairTicket).where(RepairTicket.id == ticket_id)
    result = await db.execute(stmt)
    ticket = result.scalar_one_or_none()
    if not ticket:
        return error_response("Không tìm thấy phiếu sửa chữa", 404)
    if ticket.status in ("completed", "paid"):
        return error_response("Không thể xóa phiếu đã hoàn thành hoặc đã thanh toán!", 400)
    await db.delete(ticket)
    await db.commit()
    return success_response(None, "Xóa thành công")


# ── Toggle repair item completion ──────────────────────────────────────────────

@router.put("/{ticket_id}/items/{item_id}/toggle", summary="Toggle completion status of a repair item")
async def toggle_item(
    ticket_id: int,
    item_id: int,
    body: RepairItemToggle,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(require_role(Role.ADMIN, Role.MECHANIC)),
):
    # Mechanic: verify assignment
    if current_user["role"] == Role.MECHANIC.value:
        t_stmt = select(RepairTicket).where(RepairTicket.id == ticket_id)
        t_result = await db.execute(t_stmt)
        ticket = t_result.scalar_one_or_none()
        if not ticket or ticket.mechanicName != current_user["fullName"]:
            return error_response("Bạn không được phân công phiếu này", 403)

    stmt = select(RepairItem).where(
        RepairItem.id == item_id, RepairItem.repairTicketId == ticket_id
    )
    result = await db.execute(stmt)
    item = result.scalar_one_or_none()
    if not item:
        return error_response("Không tìm thấy hạng mục", 404)

    item.isCompleted = body.isCompleted
    item.completedAt = datetime.now(timezone.utc) if body.isCompleted else None
    await db.commit()
    await db.refresh(item)
    return success_response(
        {"id": item.id, "isCompleted": item.isCompleted, "completedAt": str(item.completedAt)},
        "Đã cập nhật trạng thái hạng mục",
    )


# ── Can-complete check ─────────────────────────────────────────────────────────

@router.get("/{ticket_id}/can-complete", summary="Check if all items are done (admin + mechanic)")
async def can_complete(
    ticket_id: int,
    db: AsyncSession = Depends(get_db),
    _: dict = Depends(require_role(Role.ADMIN, Role.MECHANIC)),
):
    stmt = select(RepairTicket).where(RepairTicket.id == ticket_id)
    result = await db.execute(stmt)
    ticket = result.scalar_one_or_none()
    if not ticket:
        return error_response("Không tìm thấy phiếu sửa chữa", 404)

    total = len(ticket.items)
    done = sum(1 for i in ticket.items if i.isCompleted)
    return success_response({
        "canComplete": total > 0 and done == total,
        "totalItems": total,
        "completedItems": done,
        "incompleteItems": total - done,
    })
